### PR TITLE
NumberField fixes

### LINKED ISSUE #1486: Numberfield further work
# 🙋 Feature Request

- [x] Snapping to steps should work the same as a native spin button (hopefully allows us to combine snapValueToStep and roundToStep)
- [x] aria-description should just read "Number Field" https://github.com/adobe/react-spectrum/pull/1029#discussion_r558649145
- [x] validate after parsing https://github.com/adobe/react-spectrum/pull/1029#discussion_r553532177
- [x] followup on this comment https://github.com/adobe/react-spectrum/pull/1029#discussion_r558648042
- [x] followup on this comment https://github.com/adobe/react-spectrum/pull/1029#discussion_r542985682
- [x] followup on this comment https://github.com/adobe/react-spectrum/pull/1029#discussion_r537859943
- [x] Undo should work (try typing "123abc" and then hit cmd+z several times)
- [x] Double pixel border between input and buttons should be a single border
- [x] Focus style around disabled stepper button, disappears when scrolling and is disabled then reappears when the mouse is moved off the control. It should keep the focus style. Also Readonly focus style is wrong on hover.
- [x] Being at the max value, with the stepper up being disabled, deleting the field will not re-enable the stepper, which should start from wherever it starts as an empty input
- [x] Step 3 story can be scrolled out of range to 21
- [x] Scrolling currency switcher in Arabic with SAR accounting causes huge leaps in numbers and eventually gets stuck, both scrolling directions do the same thing

### PR DESCRIPTION
**Pull request by devongovett**
Closes #1486

* Fixes double borders between input and buttons
* Fixes disabled focus and hover states for stepper buttons
* Updates the increment/decrement button disabled states as you type rather than on blur
* Adjusts the behavior of the `step` prop to match slider, and native number input behavior more closely
* Disables the stepper buttons when adding the step value would be greater than the min/max rather than only if the current value is already outside the valid range.
* Simplify `aria-roledescription` text

## Step behavior details

* If there is a defined minimum, steps are calculated starting from the min and going upwards. For example, with `minValue={2}` and `step={3}`, the steps would be 2, 5, 8, etc.
* If there is no defined minimum, steps are calculated starting from zero, going in both directions. In other words, values are divisible by the step. For example, with `step={3}`, valid values would be -6, -3, 0, 3, 6, etc.
* When incrementing or decrementing starting from a typed value, the value changes to the nearest valid step value in the direction of travel.
* When incrementing starting from an empty field, and there is a defined minimum value, the value of the field becomes the minimum. This is an intentional deviation from native `<input type="number">` behavior, which seems to always try to start from zero even if zero is out of range.
* When decrementing starting from an empty field, and there is a defined maximum value, the value of the field becomes the maximum.  This is an intentional deviation from native `<input type="number">` behavior, which seems to always try to start from zero even if zero is out of range.
* When incrementing or decrementing starting from an empty field, and there is no defined minimum/maximum value, the value of the field becomes `0`. This is an intentional deviation from native `<input type="number">` behavior, which starts from zero but also increments/decrements, so really starts at 1 or -1.
* When committing a typed input value, if there is no step defined, we simply clamp to the min/max if any. Otherwise, we find the nearest step value.

### CODE REVIEW

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 1]
```diff
@@ -190,11 +200,13 @@ export function useNumberFieldState(
     incrementToMax,
     decrement,
     decrementToMin,
+    canIncrement,
+    canDecrement,
     minValue,
     maxValue,
-    numberValue,
+    numberValue: parsedValue,
```
Comment (snowystinger) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:207]: just to make sure I'm reading it right, the numberValue will not be the same value as someone might have from listening to onChange?

**Review by snowystinger — COMMENTED** [review]
Generally I like the direction of the changes.

I think there's still a subtle difference between native and this
try pressing up arrow on `<input type="number" min="-5" step="3"/>`, it starts at `1`
so I think the extra rule is
When the min is less than zero, start the increment from 0 but advance to the next valid step with min as our starting point.
Here's the fiddle I used to check them all, two of each line so I could easily hit both up and down without reloading https://jsfiddle.net/snowystinger/4ctejfLd/14/
The other one I cannot figure out is `<input type="number" min="-5" step="3"/>` which starts at -6 if you hit increment, but starts at -5 if you decrement.

That said, I don't know how important it is?

**Comment by devongovett** [comment]
Yes. See the PR description. These were intentional deviations.

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts (hunk 1 above)
Comment (devongovett) [reply packages/@react-stately/numberfield/src/useNumberFieldState.ts:207]: that is correct

File: packages/@react-stately/color/src/useHexColorFieldState.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ import {useColor} from './useColor';
 import {useControlledState} from '@react-stately/utils';
 import {useEffect, useState} from 'react';
 
-export interface HexColorFieldState extends Omit<NumberFieldState, 'numberValue' | 'validate' | 'minValue' | 'maxValue'> {
+export interface HexColorFieldState extends Omit<NumberFieldState, 'numberValue' | 'validate' | 'minValue' | 'maxValue' | 'canIncrement' | 'canDecrement'> {
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-stately/color/src/useHexColorFieldState.ts:20]: 🤷 maybe not, but feels like a separate refactor

File: packages/@react-spectrum/numberfield/test/NumberField.test.js [hunk 3]
```diff
@@ -338,7 +338,7 @@ describe('NumberField', function () {
     Name
     ${'NumberField'}
   `('$Name increment value by one when increment button is pressed', () => {
-    let {incrementButton} = renderNumberField({onChange: onChangeSpy});
+    let {incrementButton} = renderNumberField({defaultValue: 0, onChange: onChangeSpy});
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-spectrum/numberfield/test/NumberField.test.js:341]: yup, based on changes in pr description

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 4]
```diff
@@ -33,22 +35,20 @@ interface NumberFieldStateProps extends NumberFieldProps {
   locale: string
 }
 
-// for two decimal points of precision
-const MAX_SAFE_FLOAT = (Number.MAX_SAFE_INTEGER + 1) / 128 - 1;
-const MIN_SAFE_FLOAT = (Number.MIN_SAFE_INTEGER - 1) / 128 + 1;
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-stately/numberfield/src/useNumberFieldState.ts:38]: yeah IMO it was a bit weird to jump to these super large values

**Review by dannify — APPROVED** [review]

**Pull request merged**